In [31]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [32]:
# Load the given datasets
categorical = pd.read_csv('chronic_kidney_disease_categorical.csv')
numerical = pd.read_csv('chronic_kidney_disease_numerical.csv')
# Print the data
print(categorical)
print(numerical)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The features are the columns of our tables that correspond to different conditions wihthin the human body, and the labels indicate either the quantity in numbers or the category the feature falls in by using numerical and categorical data labels. The numerical data gives assigments purely numerical value to a specfic column to denote tests we can quantify with numbers such as the mgs/dl of Blood Glucose Random, or the Potassium in mEq/L. This is compared to the categorical data that gives a direct indication of the condition by putting it into a category based on the assessment. These are called nominal because they are represented as names within the csv or table. For example we have Albumin(nominal)	 which is represented as al  and can take on any of the values between (0,1,2,3,4,5), but since we are putting it in one of these discrete categories it is categorical. Another example would be the Bacteria(nominal) which is named ba and it could only be two things which are either present, or notpresent.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [33]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicates_numerical = numerical.duplicated(subset='unique_id')
print(f"Number of duplicate rows in numerical dataset: {duplicates_numerical.sum()}")

# Drop these duplicate rows
numerical_cleaned = numerical.drop_duplicates(subset='unique_id')

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicates_categorical = categorical.duplicated(subset='unique_id')
print(f"Number of duplicate rows in categorical dataset: {duplicates_categorical.sum()}")
categorical_cleaned = categorical.drop_duplicates(subset='unique_id')

Number of duplicate rows in numerical dataset: 30
Number of duplicate rows in categorical dataset: 30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: All of the unique ids that we have are present in both datasets because we have 481 rows which is more than the 420 or 419 we had to start. These 481 rows account for the duplicate IDs we have in both sets, as well as for the ones that could be unique to either. We can also check this by running a script on these to make sure that all ids from categorial and numerical show up in our combined data set which they do. They combined rows on unique ID show the values for the id given both numerical and categorical attributes.

In [34]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
combined_dataset = pd.merge(numerical_cleaned, categorical_cleaned, on='unique_id', how='inner')

#Print the combined dataset
print("\nCombined Dataset:")
print(combined_dataset)



Combined Dataset:
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  ...   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  ...   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
385     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0  ...   
386     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0  ...   
388     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
389     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0  ...   

           pc         pcc       

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [35]:
# Calculate the percentage of rows that contain atleast one missing value
total_rows = combined_dataset.shape[0]
missing_rows = combined_dataset.isnull().any(axis=1).sum()
percentage_missing = (missing_rows / total_rows) * 100 if total_rows > 0 else 0

# Print %
print(f"Percentage of rows with at least one missing value: {percentage_missing:.2f}%")

# Drop these rows from the dataset
cleaned_dataset = combined_dataset.dropna()

# Print the Dataset
print("\nCleaned Dataset (after dropping rows with missing values):")
print(cleaned_dataset)

Percentage of rows with at least one missing value: 60.77%

Cleaned Dataset (after dropping rows with missing values):
     unique_id   age     bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
7       109053  64.0   60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
13      349892  59.0   70.0  424.0  55.0  1.7  138.0  4.5  12.6  37.0  ...   
..         ...   ...    ...    ...   ...  ...    ...  ...   ...   ...  ...   
379     486397  20.0   70.0  123.0  44.0  1.0  135.0  3.8  14.6  44.0  ...   
381     462892  36.0   80.0   85.0  16.0  1.1  142.0  4.1  15.6  44.0  ...   
386     475200  64.0  100.0  163.0  54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0  44.0  1.2  142.0  4.9  14.5  44.0  ...   
388     474407  21.0   

## Part 3.4: Sort the dataset according to the Labels

In [36]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_dataset = cleaned_dataset.sort_values(by='Target').reset_index(drop=True)

# Print the dataset
print("\nSorted Dataset:")
print(sorted_dataset)


Sorted Dataset:
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2       484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

           pc         pcc         

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [37]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encoded_dataset = pd.get_dummies(sorted_dataset, columns=categorical_columns, dtype=int)

# Print the dataset
print("\nEncoded Dataset:")
print(encoded_dataset)


Encoded Dataset:
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2       484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

     cad_no  cad_yes  appet_good 

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [38]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

for col in numerical_columns:
    mean = encoded_dataset[col].mean()
    std_dev = encoded_dataset[col].std()
    
    # Define the outlier boundaries
    lower_bound = mean - 3 * std_dev
    upper_bound = mean + 3 * std_dev
    
    # Filter the DataFrame to include only non-outlier rows
    encoded_dataset = encoded_dataset[(encoded_dataset[col] >= lower_bound) & (encoded_dataset[col] <= upper_bound)]

# Print the dataset
print("\nCleaned Encoded Dataset without Outliers:")
print(encoded_dataset)


Cleaned Encoded Dataset without Outliers:
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
6       546225  46.0   60.0  163.0   92.0  3.3  141.0  4.0   9.8  28.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

     cad

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [39]:
from sklearn.preprocessing import MinMaxScaler

# Define the numerical columns
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

# Initialize the MinMaxScaler
scaler = MinMaxScaler()

# Normalize the numerical columns
encoded_dataset[numerical_columns] = scaler.fit_transform(encoded_dataset[numerical_columns])

# Print the normalized dataset
print("\nNormalized Encoded Dataset:")
print(encoded_dataset)


Normalized Encoded Dataset:
     unique_id       age    bp       bgr        bu        sc       sod  \
0       474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
1       137148  0.905405  1.00  0.965665  0.522876  0.641975  0.666667   
3       343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
4       532520  0.729730  0.75  0.150215  0.281046  0.234568  0.533333   
6       546225  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
..         ...       ...   ...       ...       ...       ...       ...   
148     647552  0.743243  0.25  0.214592  0.124183  0.037037  0.566667   
149     125587  0.418919  0.00  0.167382  0.241830  0.086420  0.700000   
150     614376  0.689189  0.50  0.210300  0.045752  0.098765  0.500000   
151     853426  0.472973  0.25  0.236052  0.183007  0.024691  0.666667   
152     572140  0.662162  0.50  0.145923  0.117647  0.061728  0.733333   

          pot      hemo       pcv  ...  cad_no  cad_yes  appet_good  \
0    0.2068

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: Yes, we dont need the unique_ID for modeling and predictions because they give us no insight to anything about the patient expect for identifying them from another. They only give us a way to distinugsh unique patients so using that column to predict would be unjust.

In [40]:
# Drop the identified column
encoded_dataset.drop(columns='unique_id', inplace=True)

# Print the cleaned dataset
print("\nCleaned Encoded Dataset after Removing Inappropriate Columns:")
print(encoded_dataset)
print(encoded_dataset.columns)


Cleaned Encoded Dataset after Removing Inappropriate Columns:
          age    bp       bgr        bu        sc       sod       pot  \
0    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
1    0.905405  1.00  0.965665  0.522876  0.641975  0.666667  0.000000   
3    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
4    0.729730  0.75  0.150215  0.281046  0.234568  0.533333  0.793103   
6    0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.379310   
..        ...   ...       ...       ...       ...       ...       ...   
148  0.743243  0.25  0.214592  0.124183  0.037037  0.566667  0.206897   
149  0.418919  0.00  0.167382  0.241830  0.086420  0.700000  0.689655   
150  0.689189  0.50  0.210300  0.045752  0.098765  0.500000  0.620690   
151  0.472973  0.25  0.236052  0.183007  0.024691  0.666667  0.724138   
152  0.662162  0.50  0.145923  0.117647  0.061728  0.733333  0.655172   

         hemo       pcv      wbcc  ...  cad_no  cad_yes  app

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [41]:
# Export the dataframe to a new csv file
encoded_dataset.to_csv('cleaned_encoded_dataset.csv', index=False)

# Export the dataframe to a new json file
encoded_dataset.to_json('cleaned_encoded_dataset.json', orient='records', lines=True)  # Using records format for a list of JSON objects

# Print confirmation
print(f"DataFrame has been exported to cleaned_encoded_dataset.csv and cleaned_encoded_dataset.json")

DataFrame has been exported to cleaned_encoded_dataset.csv and cleaned_encoded_dataset.json


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow


### Paste the markdown table here
|    |   age |    bp |      bgr |        bu |        sc |       sod |       pot |       hemo |       pcv |       wbcc | ... |   ba_present |   htn_yes |   dm_no |   dm_yes |   cad_yes |   appet_poor |   pe_yes |   ane_yes |   Target_ckd\t |   Target_notckd |
|----|-------|-------|----------|-----------|-----------|-----------|-----------|------------|-----------|------------|-----|--------------|------------|---------|----------|-----------|----------------|----------|-----------|------------------|------------------|
|  0 | 0.75  | 0.25  | 0.464516 | 0.562092  | 0.513043  | 0.000000  | 0.358974  | 0.311475   | 0.315789  | 0.596899   | ... |            1 |          1 |       0 |        1 |         1 |              0 |        1 |         1 |                0 |                0 |
|  8 | 0.654 | 0.75  | 0.287097 | 0.189542  | 0.095652  | 0.433333  | 0.615385  | 0.467213   | 0.473684  | 0.449612   | ... |            0 |          1 |       0 |        1 |         0 |              1 |        0 |         0 |                0 |                0 |
| 11 | 0.583 | 1.00  | 0.732258 | 0.281046  | 0.208696  | 0.633333  | 0.512821  | 0.459016   | 0.473684  | 0.759690   | ... |            0 |          1 |       0 |        1 |         0 |              0 |        0 |         0 |                0 |                0 |
| 14 | 0.643 | 1.00  | 0.622581 | 0.111111  | 0.078261  | 0.500000  | 0.461538  | 0.581967   | 0.552632  | 0.589147   | ... |            0 |          1 |       0 |        1 |         1 |              0 |        0 |         0 |                0 |                0 |
| 15 | 0.679 | 1.00  | 0.677419 | 0.163399  | 0.243478  | 0.766667  | 0.256410  | 0.606557   | 0.631579  | 0.465116   | ... |            1 |          1 |       1 |        0 |         1 |              0 |        0 |         0 |                0 |                0 |
| 21 | 0.774 | 0.00  | 0.154839 | 0.751634  | 0.426087  | 0.533333  | 0.615385  | 0.475410   | 0.500000  | 0.883721   | ... |            0 |          1 |       0 |        1 |         0 |              1 |        1 |         0 |                0 |                0 |
| 22 | 0.774 | 0.25  | 0.480645 | 0.470588  | 0.278261  | 0.433333  | 0.487179  | 0.393443   | 0.447368  | 0.139535   | ... |            1 |          1 |       0 |        1 |         1 |              0 |        0 |         0 |                0 |                0 |
| 27 | 0.583 | 0.50  | 0.464516 | 0.411765  | 0.304348  | 0.566667  | 0.615385  | 0.434426   | 0.473684  | 0.279070   | ... |            1 |          1 |       0 |        1 |         0 |              0 |        1 |         0 |                0 |                0 |
| 28 | 0.655 | 0.50  | 0.935484 | 0.058824  | 0.026087  | 0.566667  | 0.487179  | 0.786885   | 0.736842  | 0.348837   | ... |            0 |          1 |       0 |        1 |         0 |              0 |        0 |         0 |                0 |                0 |
| 32 | 0.798 | 0.50  | 0.590323 | 0.862745  | 0.365217  | 0.600000  | 0.846154  | 0.401639   | 0.447368  | 0.263566   | ... |            0 |          1 |       0 |        1 |         1 |              0 |        0 |         0 |                0 |                0 |
| 35 | 0.702 | 0.25  | 0.074194 | 0.366013  | 0.104348  | 0.566667  | 0.512821  | 0.491803   | 0.526316  | 0.627907   | ... |            0 |          0 |       0 |        1 |         0 |              0 |        0 |         0 |                0 |                0 |
| 36 | 0.488 | 1.00  | 0.432258 | 0.124183  | 0.052174  | 0.633333  | 0.435897  | 0.336066   | 0.447368  | 0.418605   | ... |            0 |          1 |       1 |        0 |         0 |              0 |        0 |         1 |                0 |                0 |
| 39 | 0.000 | 0.00  | 0.077419 | 0.372549  | 0.052174  | 0.500000  | 0.615385  | 0.352459   | 0.368421  | 1.000000   | ... |            1 |          0 |       1 |        0 |         0 |              1 |        0 |         0 |                0 |                0 |
| 41 | 0.821 | 0.50  | 0.277419 | 0.228758  | 0.173913  | 0.666667  | 0.230769  | 0.491803   | 0.500000  | 0.503876   | ... |            0 |          1 |       0 |        1 |         0 |              1 |        0 |         0 |                0 |                0 |
| 47 | 0.571 | 0.25  | 0.132258 | 0.888889  | 0.617391  | 0.700000  | 0.564103  | 0.442623   | 0.500000  | 0.372093   | ... |            0 |          0 |       1 |        0 |         0 |              0 |        0 |         0 |                0 |                0 |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

You could do it in smaller portions and have them use libraires such as pandas to do this data conversion. You shouldn't let it just run non stop, you should stop, check the output yourself, and reprompt as you keep going to mitigate data conversion hallucinations. You can also have a verifier that will check that the data is the same for a certain amount of random rows/columns and once you have a large enough sample you can make sure it converted well.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [42]:
print(encoded_dataset.columns)
# Code to rename all the columns in the dataset
column_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wbcc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'ba_present': 'bacteria presence',
    'htn_yes': 'hypertension yes',
    'dm_no': 'diabetes mellitus no',
    'dm_yes': 'diabetes mellitus yes',
    'cad_yes': 'coronary artery disease yes',
    'appet_poor': 'poor appetite',
    'pe_yes': 'pedal edema yes',
    'ane_yes': 'anemia yes',
    'Target_ckd': 'target chronic kidney disease',
    'Target_notckd': 'target not chronic kidney disease',
}

# Rename the columns in the DataFrame
encoded_dataset.rename(columns=column_mapping, inplace=True)

print(encoded_dataset.columns.tolist())


Index(['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc',
       'rbcc', 'al', 'su', 'rbc_abnormal', 'rbc_normal', 'pc_abnormal',
       'pc_normal', 'pcc_notpresent', 'pcc_present', 'ba_notpresent',
       'ba_present', 'htn_no', 'htn_yes', 'dm_no', 'dm_yes', 'cad_no',
       'cad_yes', 'appet_good', 'appet_poor', 'pe_no', 'pe_yes', 'ane_no',
       'ane_yes', 'Target_ckd', 'Target_notckd'],
      dtype='object')
['age', 'blood pressure', 'blood glucose random', 'blood urea', 'serum creatinine', 'sodium', 'potassium', 'hemoglobin', 'packed cell volume', 'white blood cell count', 'rbcc', 'albumin', 'sugar', 'rbc_abnormal', 'rbc_normal', 'pc_abnormal', 'pc_normal', 'pcc_notpresent', 'pcc_present', 'ba_notpresent', 'bacteria presence', 'htn_no', 'hypertension yes', 'diabetes mellitus no', 'diabetes mellitus yes', 'cad_no', 'coronary artery disease yes', 'appet_good', 'poor appetite', 'pe_no', 'pedal edema yes', 'ane_no', 'anemia yes', 'target chronic kidney disease', 't

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [45]:
# Converted SQL to Pandas code
result = encoded_dataset.groupby('target chronic kidney disease').size().reset_index(name='count')

# Display the result
print(result)


   target chronic kidney disease  count
0                              0    111
1                              1     22


*Describe what the above code does here*

The code above gets the two categories from the Target of either 0 (not having ckd), or 1 (having ckd). It also gives the assoictaed counts next to it which are the number of people that have it ( or the numbers of 0s and 1s summed up from the Target column). It is grouped by either having it or not as 1s and 0s.
